doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+slow, however, so to run individual tests while working on a new feature, you
+can run, e.g.:
 
-Deprecating
------------
-If you need to deprecate a function or a class, use the ``@deprecated`` decorator::
+.. code-block:: console
 
-    from mne.utils import deprecated
+    $ pytest mne/tests/test_evoked.py:test_io_evoked -x --verbose
 
-    @deprecated('my_function will be deprecated in 0.XX, please use my_new_function instead.')
-    def my_function():
-       return 'foo'
+Or alternatively:
 
-If you need to deprecate a parameter, use the mne warning function.
-For example to rename a parameter from `old_param` to `new_param` you can
-use something like this::
+.. code-block:: console
+
+    $ pytest mne/tests/test_evoked.py -k test_io_evoked -x --verbose
 
-    from mne.utils import warn
+Make sure you have the testing dataset, which you can get by running this in
+a Python interpreter::
 
-    def my_function(new_param, old_param=None):
-        if old_param is not None:
-             warn('old_param is deprecated and will be replaced by new_param in 0.XX.',
-                  DeprecationWarning)
-             new_param = old_param
-        # Do what you have to do with new_param
-        return 'foo'
+    >>> mne.datasets.testing.data_path(verbose=True)  # doctest: +SKIP
 
-Add the necessary tests to ensure that the warnings are raised properly.
-You can use something like this::
 
-    def test_my_function_depracation():
-        with pytest.deprecated_call(match="my_new_function"):
-            my_deprecated_function()
+.. _build-docs:
 
-    def test_old_param_in_my_function_deprecation():
-        with pytest.deprecated_call(match="old_param .* new_param"):
-            my_deprecated_param(old_param='bar')
+Building the documentation
+^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+Our documentation (including docstrings in code files) is in
+reStructuredText_ format and is built using Sphinx_ and `Sphinx-Gallery`_.
+The easiest way to ensure that your contributions to the documentation are
+properly formatted is to follow the style guidelines on this page, imitate
+existing documentation examples, refer to the Sphinx and Sphinx-Gallery
+reference materials when unsure how to format your contributions, and build the
+docs locally to confirm that everything looks correct before submitting the
+changes in a pull request.
 
-Profiling
----------
-To learn more about profiling Python codes please see `the scikit learn profiling site <https://scikit-learn.org/stable/developers/performance.html#performance-howto>`_.
+You can build the documentation locally using `GNU Make`_ with
+:file:`doc/Makefile`. From within the :file:`doc` directory, a full
+documentation build (of the current code state) can be triggered with
 
-.. _troubleshooting:
+.. code-block:: console
 
-Submitting changes
-------------------
-Changes to code can be submitted using a standard GitHub Pull Request, as
-documented in :ref:`using_github`.
+   $ make html_dev
```

Comment:
The full build takes a long time even on fast machines so it should not be what we tell most devs to do. This command is really reserved for the very patient (or masochistic). I'd rather tell people to use `make html_dev-noplot` to ensure formatting and links are correct, or `PATTERN=whatever.py make html_dev-pattern` if they want to build a specific example.